### PR TITLE
Update the FindImplementationsForInterfaceMember

### LINKED ISSUE #28827: Go to implementation (Ctrl+F12) is broken for  interface async methods
I defined an interface with async methods in a .Net Standard 2.0 project. And two implementations in two .Net Core 2.1 projects.  
When I press Ctrl+F12 over the interface name, the  Visua Studio 'implementations' window displays the two implementations.
But when I press Ctrl+F12 over a method whose implementation has async keyword method, it doesn't find any implementation, presenting a dialog into saying &quot;This symbol has no implementations&quot;.
The expected result is that the 'implementations' window should present the two implementations of the method.

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/295266/go-to-implementation-ctrlf12-is-broken-for-interfa.html
VSTS ticketId: 649802_
_These are the original issue comments:_
(no comments)
_These are the original issue solutions:_
(no solutions)

### PR DESCRIPTION
**Pull request by ryzngard**
Update FindImplementationsForInterfaceMember to use SymbolFinder.OriginalSymbolsMatch.

SymbolFinder.OriginalSymbolsMatch has heuristics to determine if types are
the same even if coming from a different assembly. This helps in cases
where GoToImplementation needs to resolve across projects that get
types like System.Threading.Tasks.Task from different assemblies, but
are equivalent and fulfill requirements for interface implementation.

Fixes #28827

### CODE REVIEW

**Comment by ryzngard** [comment]
Looking to add tests, once I find out where and how to

**Comment by jasonmalinowski** [comment]
@ryzngard: does this also fix #24013?

**Comment by ryzngard** [comment]
@jasonmalinowski yes, it appears to fix the case for  CodeLens and `shift+f12` scenarios as well. Tested with a netstandard2.0 library with interface and implementations in a netcore2.1 app using a method that returns `void` and one that returns `System.Threading.Tasks.Task`

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs [hunk 1]
```diff
@@ -153,14 +153,20 @@ public static ITypeSymbol RemoveNullableIfPresent(this ITypeSymbol symbol)
...
+            // Try to get the compilation for the symbol we're searching for, 
+            // which can help identify matches with the call to SymbolFinder.OriginalSymbolsMatch
+            var typeSymbolProject = solution.GetProject(typeSymbolAndProjectId.ProjectId);
+            var typeSymbolCompilation = typeSymbolProject?.GetCompilationAsync(cancellationToken).WaitAndGetResult_CanCallOnBackground(cancellationToken);
+
             foreach (var constructedInterface in constructedInterfaces)
             {
                 cancellationToken.ThrowIfCancellationRequested();
                 var constructedInterfaceMember = constructedInterface.GetMembers().FirstOrDefault(m =>
-                    SymbolEquivalenceComparer.Instance.Equals(m.OriginalDefinition, originalInterfaceMember));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: i'm concerned this won't be the only case that needs to be audited.  why not just have SymbolEquivalenceComparer be updated to understand these cases.  Having two ways of asking the exact same thing seems like a bad place to be.

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs [hunk 2]
```diff
@@ -153,14 +153,20 @@ public static ITypeSymbol RemoveNullableIfPresent(this ITypeSymbol symbol)
...
+                    SymbolFinder.OriginalSymbolsMatch(m, interfaceMember, solution, typeSymbolCompilation, null, cancellationToken));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: name arg for 'null'.

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs [hunk 3]
```diff
@@ -153,14 +153,20 @@ public static ITypeSymbol RemoveNullableIfPresent(this ITypeSymbol symbol)
...
             var originalInterfaceType = interfaceMember.ContainingType.OriginalDefinition;
             var originalInterfaceMember = interfaceMember.OriginalDefinition;
+
             var constructedInterfaces = typeSymbol.AllInterfaces.Where(i =>
                 SymbolEquivalenceComparer.Instance.Equals(i.OriginalDefinition, originalInterfaceType));
 
+            // Try to get the compilation for the symbol we're searching for, 
+            // which can help identify matches with the call to SymbolFinder.OriginalSymbolsMatch
+            var typeSymbolProject = solution.GetProject(typeSymbolAndProjectId.ProjectId);
+            var typeSymbolCompilation = typeSymbolProject?.GetCompilationAsync(cancellationToken).WaitAndGetResult_CanCallOnBackground(cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: 1. What happens if we get null for `typeSymbolCompilation` here?  
2. i would prefer we actually just move to async, and thread that backwards.

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs [hunk 4]
```diff
@@ -153,14 +153,20 @@ public static ITypeSymbol RemoveNullableIfPresent(this ITypeSymbol symbol)
             // instantiations of that method.
             var originalInterfaceType = interfaceMember.ContainingType.OriginalDefinition;
             var originalInterfaceMember = interfaceMember.OriginalDefinition;
+
             var constructedInterfaces = typeSymbol.AllInterfaces.Where(i =>
                 SymbolEquivalenceComparer.Instance.Equals(i.OriginalDefinition, originalInterfaceType));
 
+            // Try to get the compilation for the symbol we're searching for, 
+            // which can help identify matches with the call to SymbolFinder.OriginalSymbolsMatch
```
Comment (genlu) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: This just describes what the code does, but not why. Please add comment with example to explain why this is necessary.

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs (hunk 3 above)
Comment (ryzngard) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: Null `typeSymbolCompilation` is handled correctly in the following call. There are a few tests that exercise this by not providing a projectId. 

I'll look at how much work it is to push Async up.

Comment (ryzngard) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: Async streams aren't available yet, so this is fine as is for now. I don't think it's worth changing to be less lazy for this one call.

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs (hunk 1 above)
Comment (ryzngard) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: Updating `SymbolEquivalenceComparer` is a larger change and less targetted for this fix. In general I would agree if i were adding a new way to compare symbols, but I'm still only utilizing an existing functionality. If the current spread of logic isn't desired, we should file a separate issue and address accordingly.

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: i'm not asking for an async-stream, just an async method.  i.e.  return a `Task<ImmutableArray<Whatever>>` here.

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: > Updating `SymbolEquivalenceComparer` is a larger change and less targetted for this fix.

I'm skeptical.  The entire purpose of SymbolEquivalenceComparer is for this sort of thing.  So, if we don't fix that issue, we're just kicking the can down the road.  i.e. users with these same projects will likely see other semantic operations fail beyond just Find-Impl.  I would prefer a broader, more correct fix here, rather than the targetted fix.

Comment (ryzngard) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: `SymbolFinder.OriginalSymbolsMatch` does end up using `SymbolEquivalenceComparer`, it just has some more heuristics around it. http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/FindSymbols/SymbolFinder_Hierarchy.cs,9defddf5fea8d771 

I guess what I'm saying is I'm not sure that logic belongs in `SymbolEquivalenceComparer`. I don't have the breadth of knowledge to make that call. If it does, I'd prefer we file a task to remove `SymbolFinder.OriginalSymbolsMatch` and exchange to `SymbolEquivalenceComparer.IgnoreAssembliesInstance.Equals`. 

It may be sufficient to have just `SymbolEquivalenceComparer.IgnoreAssembliesInstance.Equals` here, but I opted to follow other places that seemed similar and include the added checks in `SymbolFinder.OriginalSymbolsMatch`.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: > I don't have the breadth of knowledge to make that call.

I wrote `SymbolEquivalenceComparer`.  And this was def something it was intended to support :D

I guess my poitn is this.  If we are considreing these symbols from different compilations equivalent (which seems to be how the compiler would view things), then we should update our helper to properly encode the right logic that makes it agree with what the compiler would think.  I don't think that's necessarily `IgnoreAssembliesInstance`, however it might be something else.

Do you know *why* it is that `System.Runtime.dll` and `netstandard.dll` are unifying here?  Is there a type-forwarding situation going on?

--

In other words, i'd prefer we deepely understand this and put in the right *full* fix, rather than a one-off patch at one layer that will work for this one specific case, but not be sufficient for other things we may run into later.

Comment (ryzngard) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: > Do you know why it is that System.Runtime.dll and netstandard.dll are unifying here? Is there a type-forwarding situation going on?

Yes, this is type-forwarding. For netstandard from netframework (or even netframework) reference binaries are added to do type-forwarding so `netstandard.dll` types can resolve to the equivalent type. I had an easier time viewing this in netframework build outputs. One of the referenced files is `{VsInstallPath}\MSBuild\Microsoft\Microsoft.NET.Build.Extensions\net461\lib\netstandard.dll` which contains all of the type forwarding. I'm not quite sure how it's done with netcore (I didn't look heavily), but I imagine there's something similar. 

> I guess my poitn is this. If we are considreing these symbols from different compilations equivalent (which seems to be how the compiler would view things), then we should update our helper to properly encode the right logic that makes it agree with what the compiler would think. I don't think that's necessarily IgnoreAssembliesInstance, however it might be something else.

This IS that something else. Is it just that it's not part of statics exposed by `SymbolEquivalenceComparer`? 

To me this is just swapping the wrong call to the right one. If the logic isn't in the right place, then I guess moving it would be okay? I don't feel strongly one way or another. 

I think `SymbolEquivalenceComparer` does the advertised thing and works in the expected way; you compare with a requirement for same assemblies or you compare without. I'd even be willing to move this code to calling it directly. Like I said, I just chose an example from something similar and went with it because it looked like `SymbolEquivalenceComparer` + some extra checks to ensure type forwarding is considered.

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs (hunk 3 above)
Comment (ryzngard) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: Good point. There's no real reason for this to yield in a lazy manner. I'll change to `Task<ImmutableArray<>>`

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs (hunk 1 above)
Comment (ryzngard) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: I do wonder if we should just add an extension method for `ISymbol` to facilitate these type of decisions. Looking for the correct static instance is a pain. 

`ISymbol.Equals` => `SymbolEquivalenceComparer.Instance.Equals`
`ISymbol.EqualsIgnoreAssembly` => `SymbolEquivalenceComparer.IgnoreAssembliesInstance`
`ISymbol.EqualsIgnoreAssembly(other, solution, thisCompilation, otherCompilation, cancellationToken)` => `SymbolFinder.OriginalSymbolsMatch`

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: > Yes, this is type-forwarding.

We really should just be makign SymbolEquivalenceComparer understand type forwarding.  If ti's a forwarded type, then they're equivalent.

> Is it just that it's not part of statics exposed by `SymbolEquivalenceComparer`?

Right.  The statics that are there should be updaed to understand type forwarding. 

> I think `SymbolEquivalenceComparer` does the advertised thing and works in the expected way;

I disagree.  A symbol forwarded to anotehr assembly should be equivalent to the original symbol.  That's what 'equivalence' is supposed to mean here.  i.e. "does hte compiler consider these symbols interchangeable?"  In this case, the answer is yes.  We just never added hte logic for type forwarding here since we didn't need to and there weren't reports of a problem.  Now that there are, we should fix htis.

**Comment by CyrusNajmabadi** [comment]
btw, i'd be happy to chat more about this on gitter if you think that would be helpful!
